### PR TITLE
[Merged by Bors] - feat(Algebra/Category/ModuleCat): the free presheaf of modules on a presheaf of sets

### PR DESCRIPTION
**Pull request by joelriou**
This contribution was created as part of the AIM workshop "Formalizing algebraic geometry" in June 2024.

Co-authored-by: Johan Commelin <johan@commelin.net>

---

Thanks to the refactor #16667, this is a better attempt as compared to #14245.

- [x] depends on: #16667

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by jcommelin — APPROVED** [review]
LGTM, but I was involved in writing parts of this.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+variable {R} in
+/-- Given a presheaf of types `F : Cᵒᵖ ⥤ Type u`, this is the presheaf
+of modules over `R` which sends `X : Cᵒᵖ` to the free `R.obj X`-module on `F.obj X`. -/
+@[simps]
+noncomputable def freeObj (F : Cᵒᵖ ⥤ Type u) : PresheafOfModules.{u} R where
+  obj := fun X ↦ (ModuleCat.free (R.obj X)).obj (F.obj X)
+  map := fun {X Y} f ↦ ModuleCat.freeDesc
+      (fun x ↦ ModuleCat.freeMk (F.map f x))
+  map_id := by aesop
```
Comment (chrisflav) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean:41]:
```suggestion
  obj X := (ModuleCat.free (R.obj X)).obj (F.obj X)
  map {X Y} f := ModuleCat.freeDesc (fun x ↦ ModuleCat.freeMk (F.map f x))
  map_id := by aesop
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+/-- The free presheaf of modules functors `(Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R`. -/
```
Comment (chrisflav) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean:43]:
```suggestion
/-- The free presheaf of modules functor `(Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R`. -/
```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  obj := freeObj
+  map {F G} φ :=
+    { app := fun X ↦ (ModuleCat.free (R.obj X)).map (φ.app X)
+      naturality := fun {X Y} f ↦ by
+        dsimp
+        ext x
+        simp only [ModuleCat.coe_comp, Function.comp_apply, ModuleCat.freeDesc_apply,
+          ModuleCat.restrictScalars.map_apply, ModuleCat.free_map_apply]
+        congr 1
+        exact NatTrans.naturality_apply φ f x }
```
Comment (chrisflav) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean:55]:
```suggestion
        simp [FunctorToTypes.naturality] }
```

Comment (chrisflav) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean:55]: Maybe `FunctorToTypes.naturality` could be `@[simp]`?

File: Mathlib/Algebra/Category/ModuleCat/Adjunctions.lean [hunk 4]
```diff
@@ -33,22 +33,63 @@ variable [Ring R]
...
   map_id := by intros; exact Finsupp.lmapDomain_id _ _
   map_comp := by intros; exact Finsupp.lmapDomain_comp _ _ _ _
 
+variable {R}
+
+/-- Constructor for elements in the module `(free R).obj X`. -/
+noncomputable def freeMk {X : Type u} (x : X) : (free R).obj X := Finsupp.single x 1
+
+@[ext 1200]
+lemma free_hom_ext {X : Type u} {M : ModuleCat.{u} R} {f g : (free R).obj X ⟶ M}
```
Comment (chrisflav) [inline Mathlib/Algebra/Category/ModuleCat/Adjunctions.lean:48]: Is there a reason why you sometimes write `⟶` and sometimes `→` for morphisms in `Type`?

File: Mathlib/RepresentationTheory/GroupCohomology/Resolution.lean [hunk 5]
```diff
@@ -584,8 +584,7 @@ theorem forget₂ToModuleCatHomotopyEquiv_f_0_eq :
   · dsimp only [HomotopyEquiv.ofIso, compForgetAugmentedIso]
     simp only [Iso.symm_hom, eqToIso.inv, HomologicalComplex.eqToHom_f, eqToHom_refl]
   trans (linearCombination _ fun _ => (1 : k)).comp ((ModuleCat.free k).map (terminal.from _))
-  · dsimp
-    erw [Finsupp.lmapDomain_linearCombination (α := Fin 1 → G) (R := k) (α' := ⊤_ Type u)
+  · erw [Finsupp.lmapDomain_linearCombination (α := Fin 1 → G) (R := k) (α' := ⊤_ Type u)
```
Comment (chrisflav) [inline Mathlib/RepresentationTheory/GroupCohomology/Resolution.lean:587]: This seems to be an entirely unrelated change?

**Review by chrisflav — COMMENTED** [review]
This looks good to me. You need to merge master and I left some small comments.

File: Mathlib/Algebra/Category/ModuleCat/Adjunctions.lean (hunk 4 above)
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Adjunctions.lean:48]: I would usually favour `→`. I have tried to fix one of these.

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean (hunk 3 above)
Comment (joelriou) [reply Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean]: I do not have a particular opinion about this attribute in this case.

File: Mathlib/RepresentationTheory/GroupCohomology/Resolution.lean (hunk 5 above)
Comment (joelriou) [reply Mathlib/RepresentationTheory/GroupCohomology/Resolution.lean:587]: This is necessary because the `free` no longer has the `simps` attribute.

**Comment by joelriou** [comment]
Thanks @chrisflav for your suggestions!

**Review by chrisflav — APPROVED** [review]
Thanks, LGTM!

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean [hunk 6]
```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin, Joel Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf
+import Mathlib.Algebra.Category.ModuleCat.Adjunctions
+
+/-! The free presheaf of modules on a presheaf of sets
+
```
Comment (erdOne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean:10]:
```suggestion
/-! 
# The free presheaf of modules on a presheaf of sets

```

File: Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+In this file, given a presheaf of rings `R` on a category `C`,
+we construct the functor
+`PresheafOfModules.free (Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R`
```
Comment (erdOne) [inline Mathlib/Algebra/Category/ModuleCat/Presheaf/Free.lean:13]:
```suggestion
`PresheafOfModules.free : (Cᵒᵖ ⥤ Type u) ⥤ PresheafOfModules.{u} R`
```

**Comment by erdOne** [comment]
Thanks!
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**